### PR TITLE
Add `--cfg docsrs` to all rustdoc invocations

### PR DESCRIPTION
**Pull request by Urgau**
Fixes https://github.com/rust-lang/docs.rs/issues/2389

### CODE REVIEW

File: crates/metadata/lib.rs [hunk 1]
```diff
@@ -267,10 +267,15 @@ impl Metadata {
...
+        // Unconditionnaly set `--cfg docsrs` as it has become a de-facto way to
+        // distinguish docs.rs.
+        //
+        // See https://github.com/rust-lang/docs.rs/issues/2389.
+        let mut all_rustdoc_args = vec!["--cfg".into(), "docsrs".into()];
+        all_rustdoc_args.extend_from_slice(&self.rustdoc_args);
         all_rustdoc_args.extend_from_slice(rustdoc_args);
 
-        if !self.rustc_args.is_empty() || !all_rustdoc_args.is_empty() {
+        if !self.rustc_args.is_empty() || all_rustdoc_args.len() > 2 {
```
Comment (Nemo157) [inline crates/metadata/lib.rs:278]: I think we might be able to get rid of this now, it was from when `--config` was unstable, we're not adding any other unstable flags below other than `-Z` ones which shouldn't need `-Zunstable-options`.

File: templates/core/about/builds.html [hunk 2]
```diff
@@ -29,6 +29,13 @@ <h4 id="setting-a-readme"> <a href="#setting-a-readme">Setting a README</a> </h4
     </p>
 
     <h4 id="detecting-docsrs"> <a href="#detecting-docsrs">Detecting Docs.rs</a> </h4>
+    <p>
+        To recognize Docs.rs from your Rust code, you can test for the <code>docsrs</code> cfg, e.g.:
```
Comment (Nemo157) [inline templates/core/about/builds.html:33]: Should mention that this is only set in the terminal crate, not for dependencies.

File: crates/metadata/lib.rs (hunk 1 above)
Comment (Urgau) [reply crates/metadata/lib.rs]: Done.

File: templates/core/about/builds.html (hunk 2 above)
Comment (Urgau) [reply templates/core/about/builds.html:33]: The documentation for `cargo rustdoc` [says](https://doc.rust-lang.org/nightly/cargo/commands/cargo-rustdoc.html#description):

> The specified target for the current package (or package specified by -p if provided) will be documented with the specified args being passed to the **final rustdoc invocation**. Dependencies will not be documented as part of this command.

----

So maybe we should also use "final rustdoc invocation"?

```diff
-        The `docsrs` cfg only applies to the "terminal" crate (aka the last crate build by Cargo).
-        It does not applies to dependencies or workspace members, only the "terminal" crate.
+        The `docsrs` cfg only applies to the final rustdoc invocation. It does not applies
+        to dependencies (even workspace one).
```

Comment (epage) [reply templates/core/about/builds.html:33]: `cargo rustc` also has this distinction and uses terms like in `cargo help rustc`:
- "current package" (vs "all of its dependencies")
- "final compiler invocation"

(`cargo help rustdoc` seems to gloss over this)

With a quick glance, I'm not seeing this distinction made elsewhere.

rust-lang/rfcs#3310 uses the term "root crate" for "selected packages" but our existing documentation refers to the "root crate" as the crate in the non-virtual workspace manifest.

Comment (Urgau) [reply templates/core/about/builds.html:33]: Thanks for the input. It seems like there isn't (yet) a name for this.

So for the purpose documenting the `docsrs` cfg I went ahead and updated the text to:

> The `docsrs` cfg only applies to the final rustdoc invocation (i.e. the crate currently being documented). It does not apply to dependencies (including workspace ones).

**Review by Nemo157 — APPROVED** [review]
Changes look good to me, would be good to hear if @syphar can think of any reason to not do this.

**Review by syphar — APPROVED** [review]
I agree, from what I see this has become the de-facto standard, so setting it by default is a good idea. 

You _could_ see this as a breaking change, but only if 
- someone used the feature name `docsrs` for something else, or 
- the feature exists for future usage, but shouldn't be used yet. 

Also having @epage on this topic makes me lean towards seeing these as edge-cases, and merging & deploying this.

**Comment by epage** [comment]
Good thinking through those cases and I agree that those are edge enough cases that for a service like docs.rs it seems fine for moving forward.

**Pull request merged**